wp_api/src/posts.rs

```diff
@@ -0,0 +1,468 @@
+                        .collect::<Vec<String>>()
+                        .join(","),
+                ),
+            ),
+            ("sticky", self.sticky.map(|x| x.to_string())),
+        ]
+        .into_iter()
+        // Remove `None` values
+        .filter_map(|(k, opt_v)| opt_v.map(|v| (k, v)))
+    }
+}
+
+uniffi::custom_newtype!(PostId, i32);
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+pub struct PostId(pub i32);
+
+uniffi::custom_newtype!(TagId, i32);
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+pub struct TagId(pub i32);
+
+uniffi::custom_newtype!(CategoryId, i32);
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+pub struct CategoryId(pub i32);
+
+impl std::fmt::Display for PostId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, uniffi::Record, WpContextual)]
+pub struct SparsePost {
+    #[WpContext(edit, embed, view)]
+    pub id: Option<PostId>,
+    #[WpContext(edit, view)]
+    pub date: Option<String>,
+    #[WpContext(edit, view)]
+    pub date_gmt: Option<String>,
+    #[WpContext(edit, view)]
+    pub guid: Option<PostGuid>,
+    #[WpContext(edit, embed, view)]
+    pub link: Option<String>,
+    #[WpContext(edit, view)]
+    pub modified: Option<String>,
+    #[WpContext(edit, view)]
+    pub modified_gmt: Option<String>,
+    #[WpContext(edit, embed, view)]
+    pub slug: Option<String>,
+    #[WpContext(edit, view)]
+    pub status: Option<PostStatus>,
+    #[serde(rename = "type")]
+    #[WpContext(edit, embed, view)]
+    pub post_type: Option<String>,
+    #[WpContext(edit)]
+    pub password: Option<String>,
+    #[WpContext(edit)]
+    pub permalink_template: Option<String>,
+    #[WpContext(edit)]
+    pub generated_slug: Option<String>,
+    #[WpContext(edit, embed, view)]
+    pub title: Option<PostTitle>,
+    #[WpContext(edit, view)]
+    pub content: Option<PostContent>,
+    #[WpContext(edit, embed, view)]
+    pub author: Option<UserId>,
+    #[WpContext(edit, embed, view)]
+    pub excerpt: Option<PostExcerpt>,
+    #[WpContext(edit, embed, view)]
+    pub featured_media: Option<i64>,
+    #[WpContext(edit, view)]
+    pub comment_status: Option<PostCommentStatus>,
+    #[WpContext(edit, view)]
+    pub ping_status: Option<PostPingStatus>,
+    #[WpContext(edit, view)]
+    pub format: Option<PostFormat>,
+    #[WpContext(edit, view)]
+    pub meta: Option<PostMeta>,
+    #[WpContext(edit, view)]
+    pub sticky: Option<bool>,
+    #[WpContext(edit, view)]
+    pub template: Option<String>,
+    #[WpContext(edit, view)]
+    pub categories: Option<Vec<i64>>,
+    #[WpContext(edit, view)]
+    pub tags: Option<Vec<TagId>>,
+}
+
+#[derive(Debug, Serialize, Deserialize, uniffi::Record)]
+pub struct PostGuid {
+    pub rendered: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, uniffi::Record)]
+pub struct PostTitle {
```

Comment:
These (ie – all fields related to posts) will probably need to be specialized for the context – the title object has different fields depending on the context.

It _may_ make sense to have this be an enum with each case based on the context, then those cases define the available params? Just a suggestion – feel free to disregard

Reply:
We have `#[WpContext({available_contexts})]` attributes for this, so an `enum` shouldn't be necessary. I'll see what the differences are, but for now integration tests are able to parse it for all 3 contexts 🤔